src/Diagnostics.RuntimeHost/Utilities/RuntimeLogger.cs

```diff
@@ -0,0 +1,242 @@
+            {
+                this.Provider.ScopeProvider.ForEachScope((obj, loggingProps) =>
+                {
+                    if (entry.Scopes == null)
+                    {
+                        entry.Scopes = new List<RuntimeLogScope>();
+                    }
+
+                    var scope = new RuntimeLogScope();
+                    entry.Scopes.Add(scope);
+
+                    if (obj is string)
+                    {
+                        scope.Text = obj.ToString();
+                    }
+                    else if (obj is IEnumerable<KeyValuePair<string, object>> properties)
+                    {
+                        if (scope.Properties == null)
+                        {
+                            scope.Properties = new Dictionary<string, object>();
+                        }
+
+                        foreach (var item in properties)
+                        {
+                            scope.Properties[item.Key] = item.Value;
+                        }
+                    }
+                }, state);
+            }
+
+            this.Provider.WriteLog(entry);
+        }
+    }
+
+    [Microsoft.Extensions.Logging.ProviderAlias("Runtime")]
+    public class RuntimeLogProvider : IDisposable, IRuntimeLoggerProvider, ISupportExternalScope
+    {
+        private ConcurrentDictionary<string, RuntimeLogger> _loggers = new ConcurrentDictionary<string, RuntimeLogger>();
+        private IExternalScopeProvider _scopeProvider;
+        public LogLevel LogLevel { get; set; } = LogLevel.Information;
+        private IDisposable _settingsChangeToken;
+        private bool _disposed = false;
+        private static ConcurrentDictionary<string, List<RuntimeEventLogEntry>> _runtimeLogs = new ConcurrentDictionary<string, List<RuntimeEventLogEntry>>();
+
+        public RuntimeLogProvider()
+        {
+        }
+
+        ~RuntimeLogProvider()
+        {
+            if (!_disposed)
+            {
+                Dispose(false);
+            }
+        }
+
+        internal IExternalScopeProvider ScopeProvider
+        {
+            get
+            {
+                if (_scopeProvider == null)
+                {
+                    _scopeProvider = new LoggerExternalScopeProvider();
+                }
+
+                return _scopeProvider;
+            }
+        }
+
+        public void WriteLog(RuntimeEventLogEntry info)
+        {
+            if (info == null || info.Message == null)
+            {
+                return;
+            }
+
+            var reqid = info.Category;
+            _runtimeLogs.AddOrUpdate(reqid,
+                id => new List<RuntimeEventLogEntry> { info }, // Adding
+                (id, logs) => // Updating
+                {
+                    logs.Add(info);
+                    return logs;
+                });
+
+            // TODO: emit ETW event
+            ////DiagnosticsETWProvider.Instance.LogRuntimeHostMessage(info.Message);
```

Comment:
Should this comment be removed?

Reply:
You are right. I will just leave the TODO comment and delete the next time.

Reply:
Removed